File: engine/vastrangam/logs.py
"""Effective-dated logs — Part 2 and Part 3 of the spec.

A value is never overwritten. The open row is closed and a new one appended, so
history stays intact and a future-dated row activates by itself when that month
is reached.

Zero matches is an error, not zero. Silently returning 0 is how a person earns
nothing without anyone noticing.
"""

from __future__ import annotations

import datetime as _dt
from dataclasses import dataclass, field
from typing import Any, Iterable

from .calendar_util import Month, day_before, parse_date

FOREVER = _dt.date(9999, 12, 31)


class Unresolved(LookupError):
    """No row in this log covers that month. The month is unresolvable."""

    def __init__(self, log: str, key: str, month: Month):
        self.log, self.key, self.month = log, key, month
        super().__init__(f"{log}: nothing in force for {key!r} in {month}")


class Ambiguous(LookupError):
    """More than one row covers that month. The log contradicts itself."""

    def __init__(self, log: str, key: str, month: Month, rows):
        self.log, self.key, self.month, self.rows = log, key, month, rows
        spans = " / ".join(r.span for r in rows)
        super().__init__(f"{log}: {len(rows)} rows in force for {key!r} in {month} -> {spans}")


@dataclass(frozen=True)
class LogRow:
    key: str
    frm: _dt.date
    to: _dt.date | None
    value: Any

    @property
    def open(self) -> bool:
        return self.to is None

    @property
    def end(self) -> _dt.date:
        return self.to or FOREVER

    @property
    def span(self) -> str:
        return f"{self.frm.isoformat()}..{'open' if self.open else self.to.isoformat()}"

    def covers_month(self, month: Month) -> bool:
        return self.frm <= month.last_day and self.end >= month.first_day

    def covers_date(self, d: _dt.date) -> bool:
        return self.frm <= d <= self.end

    def days_in(self, month: Month) -> int:
        """How many days of the month this row is actually in force for."""
        lo = max(self.frm, month.first_day)
        hi = min(self.end, month.last_day)
        return max(0, (hi - lo).days + 1)


class EffectiveLog:
    """One effective-dated log. Keyed by staff id, karigar id — whatever it tracks."""

    def __init__(self, name: str):
        self.name = name
        self._rows: dict[str, list[LogRow]] = {}

    # -- writing -------------------------------------------------------------

    def set_value(self, key: str, frm, value) -> LogRow:
        """The only way policy changes. Close the open row, append the new one.

        Rows that already ended are never touched. A from-date earlier than the
        open row's start would rewrite history, so it is refused.
        """
        frm = parse_date(frm)
        if frm is None:
            raise ValueError(f"{self.name}: set_value needs a from-date for {key!r}")
        rows = self._rows.setdefault(key, [])
        for i, row in enumerate(rows):
            if row.open:
                if frm <= row.frm:
                    raise ValueError(
                        f"{self.name}: {key!r} already opens at {row.frm} — "
                        f"cannot start a new row at {frm} without rewriting history"
                    )
                rows[i] = LogRow(key, row.frm, day_before(frm), row.value)
        row = LogRow(key, frm, None, value)
        rows.append(row)
        rows.sort(key=lambda r: r.frm)
        return row

    def add(self, key: str, frm, to, value) -> LogRow:
        """Load a closed row verbatim. Used by importers replaying known history."""
        row = LogRow(key, parse_date(frm), parse_date(to), value)
        if row.to and row.to < row.frm:
            raise ValueError(f"{self.name}: {key!r} ends {row.to} before it starts {row.frm}")
        rows = self._rows.setdefault(key, [])
        rows.append(row)
        rows.sort(key=lambda r: r.frm)
        return row

    def load(self, entries: Iterable[dict]) -> "EffectiveLog":
        """Bulk load. Each entry is {key, from, to?, value}."""
        for e in entries:
            self.add(e["key"], e["from"], e.get("to"), e["value"])
        return self

    # -- reading -------------------------------------------------------------

    def keys(self):
        return list(self._rows)

    def rows(self, key: str | None = None) -> list[LogRow]:
        if key is None:
            return [r for rs in self._rows.values() for r in rs]
        return list(self._rows.get(key, ()))

    def overlapping(self, key: str, month) -> list[LogRow]:
        month = Month.of(month)
        return [r for r in self._rows.get(key, ()) if r.covers_month(month)]

    def resolve(self, key: str, month) -> Any:
        """The value in force for that month. Exactly one row must match."""
        return self.resolve_row(key, month).value

    def resolve_row(self, key: str, month) -> LogRow:
        month = Month.of(month)
        hits = self.overlapping(key, month)
        if not hits:
            raise Unresolved(self.name, key, month)
        if len(hits) > 1:
            raise Ambiguous(self.name, key, month, hits)
        return hits[0]

    def maybe(self, key: str, month, default=None):
        """For genuinely optional logs only — a piece rate for a salaried person.

        Never use this to paper over a missing salary or threshold. A missing
        mandatory value must reach the caller as Unresolved.
        """
        try:
            return self.resolve(key, month)
        except Unresolved:
            return default

    def segments(self, key: str, month) -> list[tuple[LogRow, int]]:
        """Every row touching the month with the days it was in force.

        `resolve` demands one row because pay is a monthly figure. When a value
        genuinely changes mid-month the log is ambiguous by design — this is how
        a report shows why, and how proration could be added without guessing.
        """
        month = Month.of(month)
        return [(r, r.days_in(month)) for r in self.overlapping(key, month)]

    def on(self, key: str, when) -> Any:
        """The value in force on a single date. Used by daily rules, not pay."""
        d = parse_date(when)
        for r in self._rows.get(key, ()):
            if r.covers_date(d):
                return r.value
        raise Unresolved(self.name, key, Month.of(d))

    def to_json(self) -> list[dict]:
        out = []
        for r in sorted(self.rows(), key=lambda r: (r.key, r.frm)):
            out.append(
                {
                    "key": r.key,
                    "from": r.frm.isoformat(),
                    "to": r.to.isoformat() if r.to else None,
                    "value": r.value,
                }
            )
        return out

    def __len__(self):
        return sum(len(v) for v in self._rows.values())

    def __repr__(self):
        return f"<EffectiveLog {self.name}: {len(self)} rows, {len(self._rows)} keys>"


@dataclass
class Spell:
    """One period of employment. A person may have many — they leave and return."""

    key: str
    joined: _dt.date
    left: _dt.date | None = None

    def covers(self, month) -> bool:
        month = Month.of(month)
        end = self.left or _dt.date(9999, 12, 31)
        return self.joined <= month.last_day and end >= month.first_day

    @property
    def span(self) -> str:
        return f"{self.joined.isoformat()}..{'open' if not self.left else self.left.isoformat()}"


class SpellLog:
    """Employment spells. Deliberately not an EffectiveLog: gaps are meaningful,
    overlaps are not, and 'not employed' is an answer rather than an error."""

    def __init__(self, name: str = "employment"):
        self.name = name
        self._spells: dict[str, list[Spell]] = {}

    def join(self, key: str, joined, left=None) -> Spell:
        s = Spell(key, parse_date(joined), parse_date(left))
        if s.left and s.left < s.joined:
            raise ValueError(f"{key!r} left {s.left} before joining {s.joined}")
        for other in self._spells.get(key, ()):
            if _overlap(s, other):
                raise ValueError(f"{key!r} has overlapping spells {other.span} and {s.span}")
        self._spells.setdefault(key, []).append(s)
        self._spells[key].sort(key=lambda x: x.joined)
        return s

    def leave(self, key: str, left) -> Spell:
        """Close the open spell. Returning later is a new spell, not an edit."""
        left = parse_date(left)
        for s in self._spells.get(key, ()):
            if s.left is None:
                s.left = left
                return s
        raise LookupError(f"{key!r} has no open spell to close")

    def spells(self, key: str) -> list[Spell]:
        return list(self._spells.get(key, ()))

    def keys(self):
        return list(self._spells)

    def employed(self, key: str, month) -> bool:
        return any(s.covers(month) for s in self._spells.get(key, ()))

    def employed_months(self, key: str, months) -> list[Month]:
        return [m for m in months if self.employed(key, m)]

    def to_json(self) -> list[dict]:
        return [
            {"key": s.key, "joined": s.joined.isoformat(),
             "left": s.left.isoformat() if s.left else None}
            for k in sorted(self._spells)
            for s in self._spells[k]
        ]

    def __len__(self):
        return sum(len(v) for v in self._spells.values())


def _overlap(a: Spell, b: Spell) -> bool:
    a_end = a.left or _dt.date(9999, 12, 31)
    b_end = b.left or _dt.date(9999, 12, 31)
    return a.joined <= b_end and b.joined <= a_end

File: core/logs.js
'use strict';
/* Effective-dated logs — ported rule for rule from the tenant engine's Python
 * implementation, which is where the behaviour below was worked out and is tested
 * against a real payroll. That file is not in this archive: it belongs to a tenant, and
 * the product ships without one. Naming its path here sent a reader to a file they do
 * not have and put a customer's name in the product's own engine, which is the one
 * place §0 says it may never be.

   A value is never overwritten. The open row is closed and a new one appended,
   so history stays intact and a future-dated row activates by itself when that
   month arrives.

   The rule that matters most, and the reason this is a shared primitive rather
   than something each module reinvents:

       Zero matches is an ERROR, not zero.

   Silently returning 0 is how a person earns nothing without anyone noticing,
   how a tax rate becomes 0% for a month, and how a price becomes free. Every
   log in the system — salary, threshold, piece rate, GST rate, commission,
   channel price, credit limit — resolves through here.

   The accounting prompt asks for exactly this shape independently (§4:
   "tax rate must be versioned by effective date, not a single static field, so
   historical invoices remain correct"), which is why it lives in core/ and not
   in the payroll module.

   Pure. No database, no DOM. */

const FOREVER = '9999-12-31';

class Unresolved extends Error {
  constructor(log, key, when) {
    super(`${log}: nothing in force for ${JSON.stringify(key)} in ${when}`);
    this.name = 'Unresolved';
    this.log = log; this.key = key; this.when = when;
  }
}

class Ambiguous extends Error {
  constructor(log, key, when, rows) {
    const spans = rows.map((r) => r.span).join(' / ');
    super(`${log}: ${rows.length} rows in force for ${JSON.stringify(key)} in ${when} -> ${spans}`);
    this.name = 'Ambiguous';
    this.log = log; this.key = key; this.when = when; this.rows = rows;
  }
}

/* -- dates ---------------------------------------------------------------
   ISO strings throughout. They compare correctly with <= and >=, they survive
   JSON without a timezone shifting them by a day, and SQLite stores them as
   text. A Date object here would be a bug waiting for a machine in another
   timezone to run payroll. */

function isoDate(value) {
  if (value == null || value === '') return null;
  if (value instanceof Date) return value.toISOString().slice(0, 10);
  const text = String(value).trim();
  let m = /^(\d{4})-(\d{2})-(\d{2})/.exec(text);
  if (m) return `${m[1]}-${m[2]}-${m[3]}`;
  m = /^(\d{1,2})[-/](\d{1,2})[-/](\d{4})$/.exec(text);           // dd-mm-yyyy
  if (m) return `${m[3]}-${m[2].padStart(2, '0')}-${m[1].padStart(2, '0')}`;
  throw new Error(`not a date: ${value}`);
}

/** '2025-04' or '2025-04-17' -> {first, last} of that month. */
function monthSpan(month) {
  const text = String(month).trim();
  const m = /^(\d{4})-(\d{1,2})/.exec(text);
  if (!m) throw new Error(`not a month: ${month}`);
  const y = Number(m[1]);
  const mo = Number(m[2]);
  const last = new Date(Date.UTC(y, mo, 0)).getUTCDate();
  const mm = String(mo).padStart(2, '0');
  return { key: `${y}-${mm}`, first: `${y}-${mm}-01`, last: `${y}-${mm}-${last}` };
}

function dayBefore(iso) {
  const d = new Date(`${isoDate(iso)}T00:00:00Z`);
  d.setUTCDate(d.getUTCDate() - 1);
  return d.toISOString().slice(0, 10);
}

function daysBetween(a, b) {
  const ms = new Date(`${b}T00:00:00Z`) - new Date(`${a}T00:00:00Z`);
  return Math.round(ms / 86400000);
}

class LogRow {
  constructor(key, frm, to, value) {
    this.key = key;
    this.frm = isoDate(frm);
    this.to = to == null ? null : isoDate(to);
    this.value = value;
    Object.freeze(this);
  }
  get open() { return this.to === null; }
  get end() { return this.to === null ? FOREVER : this.to; }
  get span() { return `${this.frm}..${this.open ? 'open' : this.to}`; }
  coversMonth(span) { return this.frm <= span.last && this.end >= span.first; }
  coversDate(d) { return this.frm <= d && d <= this.end; }
  /** How many days of the month this row is actually in force for. */
  daysIn(span) {
    const lo = this.frm > span.first ? this.frm : span.first;
    const hi = this.end < span.last ? this.end : span.last;
    return Math.max(0, daysBetween(lo, hi) + 1);
  }
}

class EffectiveLog {
  constructor(name) {
    this.name = name;
    this._rows = new Map();
  }

  // -- writing ------------------------------------------------------------

  /** The only way policy changes. Close the open row, append the new one.
   *  Rows that already ended are never touched, and a from-date at or before
   *  the open row's start is refused because it would rewrite history. */
  setValue(key, frm, value) {
    const from = isoDate(frm);
    if (!from) throw new Error(`${this.name}: setValue needs a from-date for ${key}`);
    const rows = this._rows.get(key) || [];
    for (let i = 0; i < rows.length; i++) {
      const row = rows[i];
      if (row.open) {
        if (from <= row.frm) {
          throw new Error(
            `${this.name}: ${JSON.stringify(key)} already opens at ${row.frm} — cannot ` +
            `start a new row at ${from} without rewriting history`
          );
        }
        rows[i] = new LogRow(key, row.frm, dayBefore(from), row.value);
      }
    }
    const row = new LogRow(key, from, null, value);
    rows.push(row);
    rows.sort((a, b) => (a.frm < b.frm ? -1 : a.frm > b.frm ? 1 : 0));
    this._rows.set(key, rows);
    return row;
  }

  /** Load a closed row verbatim. For importers replaying known history. */
  add(key, frm, to, value) {
    const row = new LogRow(key, frm, to, value);
    if (row.to && row.to < row.frm) {
      throw new Error(`${this.name}: ${key} ends ${row.to} before it starts ${row.frm}`);
    }
    const rows = this._rows.get(key) || [];
    rows.push(row);
    rows.sort((a, b) => (a.frm < b.frm ? -1 : a.frm > b.frm ? 1 : 0));
    this._rows.set(key, rows);
    return row;
  }

  load(entries) {
    for (const e of entries) this.add(e.key, e.from, e.to ?? null, e.value);
    return this;
  }

  // -- reading ------------------------------------------------------------

  keys() { return [...this._rows.keys()]; }

  rows(key) {
    if (key === undefined) return [...this._rows.values()].flat();
    return [...(this._rows.get(key) || [])];
  }

  overlapping(key, month) {
    const span = monthSpan(month);
    return (this._rows.get(key) || []).filter((r) => r.coversMonth(span));
  }

  /** The value in force for that month. Exactly one row must match. */
  resolve(key, month) { return this.resolveRow(key, month).value; }

  resolveRow(key, month) {
    const hits = this.overlapping(key, month);
    if (hits.length === 0) throw new Unresolved(this.name, key, monthSpan(month).key);
    if (hits.length > 1) throw new Ambiguous(this.name, key, monthSpan(month).key, hits);
    return hits[0];
  }

  /** For genuinely optional logs only — a piece rate for a salaried person.
   *  Never use this to paper over a missing salary, threshold or tax rate: a
   *  missing mandatory value must reach the caller as Unresolved. */
  maybe(key, month, fallback = null) {
    try { return this.resolve(key, month); }
    catch (e) { if (e instanceof Unresolved) return fallback; throw e; }
  }

  /** Every row touching the month with the days it was in force. resolve()
   *  demands one row because pay is a monthly figure; when a value genuinely
   *  changes mid-month the log is ambiguous by design, and this is how a report
   *  shows why without guessing at proration. */
  segments(key, month) {
    const span = monthSpan(month);
    return this.overlapping(key, month).map((r) => ({ row: r, days: r.daysIn(span) }));
  }

  /** The value in force on a single date. For daily rules — a GST rate on an
   *  invoice date, a price on an order date — not for monthly pay. */
  on(key, when) {
    const d = isoDate(when);
    for (const r of this._rows.get(key) || []) if (r.coversDate(d)) return r.value;
    throw new Unresolved(this.name, key, String(d).slice(0, 7));
  }

  toJSON() {
    return this.rows()
      .sort((a, b) => (a.key === b.key ? (a.frm < b.frm ? -1 : 1) : a.key < b.key ? -1 : 1))
      .map((r) => ({ key: r.key, from: r.frm, to: r.to, value: r.value }));
  }

  get size() { return this.rows().length; }
}

/** A spell log — employment, a channel being connected, a vendor being active.
 *  Unlike EffectiveLog it carries no value and a key may have several spells,
 *  because a person can leave and come back. */
class SpellLog {
  constructor(name) { this.name = name; this._rows = new Map(); }

  join(key, frm, to = null) {
    const rows = this._rows.get(key) || [];
    rows.push(new LogRow(key, frm, to, true));
    rows.sort((a, b) => (a.frm < b.frm ? -1 : 1));
    this._rows.set(key, rows);
    return rows[rows.length - 1];
  }

  keys() { return [...this._rows.keys()]; }
  rows(key) { return key === undefined ? [...this._rows.values()].flat() : [...(this._rows.get(key) || [])]; }

  /** Was this key active at any point during that month? */
  active(key, month) {
    const span = monthSpan(month);
    return (this._rows.get(key) || []).some((r) => r.coversMonth(span));
  }

  activeOn(key, when) {
    const d = isoDate(when);
    return (this._rows.get(key) || []).some((r) => r.coversDate(d));
  }

  toJSON() {
    return this.rows().map((r) => ({ key: r.key, from: r.frm, to: r.to }));
  }
}

module.exports = {
  FOREVER, Unresolved, Ambiguous, LogRow, EffectiveLog, SpellLog,
  isoDate, monthSpan, dayBefore, daysBetween,
};
